compiler/src/iree/compiler/Codegen/Common/VectorLayoutAnalysis.cpp

```diff
@@ -0,0 +1,931 @@
+  for (OpOperand &operand : op->getOpOperands()) {
+    if (operand.get().getType().isa<VectorType>()) {
+      if (operandIndex == operandLatticeIndex) {
+        return operand;
+      }
+      operandIndex++;
+    }
+  }
+  llvm_unreachable("No vector operand found");
+}
+
+/// Get a layout if all the given layouts are same. If all layouts are not same,
+/// return nullptr.
+static DistributionLayout *
+getAgreedLayout(ArrayRef<DistributionLayout *> layouts) {
+  if (layouts.size() == 0)
+    return nullptr;
+
+  // Check if all layouts are same.
+  for (unsigned i = 1, e = layouts.size(); i < e; ++i) {
+    if (*layouts[i] != *layouts[0]) {
+      return nullptr;
+    }
+  }
+
+  return layouts[0];
+}
+
+/// Get a layout if all the given layouts are same. If all layouts are not same,
+/// return nullptr.
+static const DistributionLayout *
+getAgreedLayout(ArrayRef<const DistributionLayout *> layouts) {
+  if (layouts.size() == 0)
+    return nullptr;
+
+  // Check if all layouts are same.
+  for (unsigned i = 1, e = layouts.size(); i < e; ++i) {
+    if (*layouts[i] != *layouts[0]) {
+      return nullptr;
+    }
+  }
+
+  return layouts[0];
+}
+
+/// Given a list of layouts, enforce a single layout for all of them.
+/// The layout chosen is a heuristic that choses the first enforced layout.
+/// TODO: Use the most common layout to minimize the number of conflicts.
+static void enforceSameLayoutForOperands(
+    Operation *op, ArrayRef<DistributionLayout *> operands,
+    std::function<void(DistributionLayout *, ChangeResult)> update) {
+  // Get any enforced layout.
+  DistributionLayout *chosenOperandLayout = nullptr;
+  for (DistributionLayout *lattice : operands) {
+    if (lattice->hasLayout()) {
+      chosenOperandLayout = lattice;
+      break;
+    }
+  }
+
+  // Enforce the layout to other operands.
+  if (chosenOperandLayout) {
+    // Note that the operand lattice is not updated. So using the operand
+    // lattice again can cause bugs.
+    for (auto [index, lattice] : llvm::enumerate(operands)) {
+      OpOperand &opOperand = getOpOperand(op, index);
+      ChangeResult changed =
+          lattice->resolveWithPossibleConflict(chosenOperandLayout, opOperand);
+      update(lattice, changed);
+    }
+  }
+}
+
+/// ==========================================================================
+///        PROPAGATION TRANSFER FUNCTIONS
+/// ==========================================================================
+
+static void propagateLayoutToElementwiseOp(
+    Operation *op, ArrayRef<const DistributionLayout *> operandLattices,
+    ArrayRef<DistributionLayout *> resultLattices,
+    std::function<void(DistributionLayout *, ChangeResult)> update) {
+  // All operands and results must agree on the same layout.
+
+  // We do not support multiple results yet.
+  if (resultLattices.size() != 1)
+    return;
+
+  DistributionLayout *result = resultLattices[0];
+
+  // If result lattice already has a layout, we cannot do
```

Comment:
This is so that we don't insert conflicts in the propagation pass. It's easier to insert conflicts during enforcement, because we know the user of the conflict.